icepack/models/shallow_ice.py

```diff
@@ -0,0 +1,187 @@
+
+def part_2(u, h, s, A):
+    r"""Return part 2 of the shallow ice action functional
+
+    The part 2 of the shallow ice action functional is
+
+    .. math::
+        E(u) = \int_\Omega\frac{2A(\varrho_I g)**n}{n+2} (\nabla h^2\cdot u) h^{n+1} \nabla s^{n-1}\; dx
+
+    Parameters
+    ----------
+    u : firedrake.Function
+        ice velocity
+    h : firedrake.Function
+        ice thickness
+    s : firedrake.Function
+        ice surface elevation
+    A : firedrake.Function
+        rate factor
+
+    Returns
+    -------
+    firedrake.Form
+    """
+    return ((2 * A * (ρ_I * g)**n) / (n + 2)) * inner(grad(s), u) * (h ** (n + 1)) * (grad(s) ** (n - 1))
+
+
+def penalty(u):
+    r"""Return the penalty of the shallow ice action functional
+
+    The penalty of the shallow ice action functional is 
+
+    .. math::
+        E(u) = \frac{1}{2}\int_\Omega l^2\nabla u\cdot \nabla u\; dx
+
+    Parameters
+    ----------
+    u : firedrake.Function
+        ice velocity
+
+    Returns
+    -------
+    firedrake.Form
+    """
+    l = 2 * firedrake.CellDiameter(u.ufl_domain())
+    return .5 * l**2 * inner(grad(u), grad(u))
+
+
+class ShallowIce(object):
+    r"""Class for modelling the flow of grounded ice
+    This class provides functions that solve for the velocity, thickness,
+    and surface elevation of a grounded area of slow flowing ice.
+
+    """
+    def __init__(self, part_1=part_1, part_2=part_2, penalty=penalty, 
+                 mass_transport=LaxWendroff()):
+        self.mass_transport = mass_transport
+        self.part_1 = add_kwarg_wrapper(part_1)
+        self.part_2 = add_kwarg_wrapper(part_2)
+        self.penalty = add_kwarg_wrapper(penalty)
+
+    def action(self, u, h, s, A, **kwargs):
+        r"""Return the action functional that gives the shallow ice
+        diagnostic model as the Euler-Lagrange equations
+
+        Parameters
+        ----------
+        u : firedrake.Function
+            Ice velocity
+        h : firedrake.Function
+            Ice thickness
+        s : firedrake.Function
+            Ice surface elevation
+        A : firedrake.
+            Rate factor
+
+        Returns
+        -------
+        E : firedrake.Function
+            the shallow ice action functional
+
+        Other parameters
+        ----------------
+        **kwargs
+            All other keyword arguments will be passed on to the
+            'part_1', 'part_2' and 'penalty' functionals
+        """
+        part_1 = self.part_1(u=u, **kwargs) * dx
+        part_2 = self.part_2(u=u, h=h, s=s, A=A, **kwargs) * dx
+        penalty = self.penalty(u=u, **kwargs) * dx
+        return part_1 + part_2 + penalty
+
+    def scale(self, u, **kwargs):
+        r"""Return the positive, convex part of the action functional
+
+        The positive part of the action functional is used as a dimensional
+        scale to determine when to terminate an optimization algorithm.
+        """
+        return self.part_1(u=u, **kwargs)
```

Comment:
I think this needs a `* dx`? Also, the penalty term in the action is strictly positive, so we could include that in the scale functional as well.

Reply:
Somehow I missed this comment in my initial changes. A new commit is coming. 
I moved the * dx to the action function as I think we had discussed. Do you want me to leave the * dx in the old location (as part of the mass, gravity, and penalty functions, as you have indicated here) or in the new location (where it currently is in the action function)? In the old location, it will be consistent with the rest of the models, but in the new location it will be consistent with changes you are about to make.

Reply:
I added the penalty term to the scale, but I also realized that scale is not used in the shallow ice model. Should it be kept for future use or discarded?

Reply:
The new location for all the measures is good. Good catch about the scale functional -- I think we should keep it for now. We can always remove it later. It might come in handy if, say, you want to add sliding.

Reply:
I'm surprised the coverage checker hasn't chimed in yet, but if we do keep it, we should call the scale functional somewhere in the shallow ice test -- something as simple as

    assert firedrake.assemble(shallow_ice.scale(u=u, h=h, s=s, A=A) > 0)

would be plenty.

Reply:
I added the penalty term to the scale function and included a test of the scale (ensuring that it is indeed positive) to the shallow_ice_test.py script.